generator/generator.go

```diff
@@ -196,13 +200,84 @@ func makeOutputSpec(goCode *ygen.GeneratedGoCode) map[string]codeOut {
+// the code to be output to that filename. It allows division of a
+// ygen.GeneratedGoCode struct into a set of source files. It divides the
+// methods, interfaces, and enumeration code snippets into their own files.
+// Structs are output into files by splitting them evenly among the input split
+// number.  first letter of their name within the code.
+func splitCodeByFileN(goCode *ygen.GeneratedGoCode, fileN int) (map[string]codeOut, error) {
+	structN := len(goCode.Structs)
+	if fileN < 1 || fileN > structN {
+		return nil, fmt.Errorf("requested %d files, but must be between 1 and %d (number of schema structs)", fileN, structN)
+	}
+
+	var structFiles []codeOut
+	var code, methodCode, interfaceCode strings.Builder
+	structsPerFile := int(math.Ceil(float64(structN) / float64(fileN)))
+	// Empty files could appear with certain structN/fileN combinations due
+	// to the ceiling numbers being used for structsPerFile.
+	// e.g. 4/3 gives two files of two structs.
+	// This is a little more complex, but spreads out the structs more evenly.
+	// If we instead use the floor number, and put all remainder structs in
+	// the last file, we might double the last file's number of structs if we get unlucky.
+	// e.g. 99/10 assigns 18 structs to the last file.
+	emptyFiles := fileN - int(math.Ceil(float64(structN)/float64(structsPerFile)))
+	for i, s := range goCode.Structs {
+		code.WriteString(s.StructDef)
+		code.WriteString(s.ListKeys)
+		code.WriteString("\n")
+		methodCode.WriteString(s.Methods)
+		methodCode.WriteString("\n")
+		interfaceCode.WriteString(s.Interfaces)
+		interfaceCode.WriteString("\n")
+		// The last file contains the remainder of the structs.
+		if i == structN-1 || (i+1)%structsPerFile == 0 {
+			structFiles = append(structFiles, codeOut{contents: code.String()})
+			code.Reset()
+		}
+	}
+	for i := 0; i != emptyFiles; i++ {
+		structFiles = append(structFiles, codeOut{contents: "", writeEvenIfEmpty: true})
+	}
+
+	code.Reset()
+	code.WriteString(goCode.EnumMap)
+	if code.Len() != 0 {
+		code.WriteString("\n")
+	}
+	code.WriteString(goCode.EnumTypeMap)
+
+	out := map[string]codeOut{
+		enumMapFn:   {contents: code.String()},
+		schemaFn:    {contents: goCode.JSONSchemaCode},
+		interfaceFn: {contents: interfaceCode.String()},
+		methodFn:    {contents: methodCode.String(), oneoffHeader: true},
+		enumFn:      {contents: strings.Join(goCode.Enums, "\n")},
+	}
+
+	for i, structFile := range structFiles {
+		out[fmt.Sprintf("%s%d.go", structBaseFn, i)] = structFile
+	}
+
+	return out, nil
+}
+
 // writeGoCodeMultipleFiles writes the input goCode to a set of files as specified
 // by specification returned by output spec.
-func writeGoCodeMultipleFiles(dir string, goCode *ygen.GeneratedGoCode) error {
-	out := makeOutputSpec(goCode)
+func writeGoCodeMultipleFiles(dir string, goCode *ygen.GeneratedGoCode, fileN int) error {
+	var out map[string]codeOut
+	if fileN == 0 {
+		out = splitCodeByLeadingLetter(goCode)
```

Comment:
Shall we just deprecate this option? I don't know of anyone that is using it - and it seems to be more problematic to make work with build systems that expect a specific number of files to exist.

Reply:
Removed it. I also removed the `methods.go` file and merged them into the `structs-%d.go` files, because those also scale up pretty quickly. I merged them because I thought it might make sense to put the methods alongside the structs. However, I also see how splitting them might make the structs file easier to browse, so I can put them into a `methods-%d.go` file if the latter makes more sense.

Reply:
I think this makes sense - the only real advantage of this was that I thought that folks wouldn't really want to look at the methods, however, I'm not sure it ever really made entire sense. Let's go with this approach and we can always re-do this if we want to in the future.